src/Nethermind/Nethermind.State/PreBlockCaches.cs

```diff
@@ -5,28 +5,101 @@
-    public ConcurrentDictionary<AddressAsKey, Account> StateCache { get; } = new(Environment.ProcessorCount * 2, 4096 * 4);
-    public ConcurrentDictionary<NodeKey, byte[]?> RlpCache { get; } = new(Environment.ProcessorCount * 2, 4096 * 4);
-    public ConcurrentDictionary<PrecompileCacheKey, (ReadOnlyMemory<byte>, bool)> PrecompileCache { get; } = new(Environment.ProcessorCount * 2, 4096 * 4);
+    private const int InitialCapacity = 4096 * 8;
+    private static int LockPartitions => CollectionExtensions.LockPartitions;
+
+    private readonly ILogger _logger;
+    private readonly Action _clearStorageCache;
+    private readonly Action _clearStateCache;
+    private readonly Action _clearRlpCache;
+    private readonly Action _clearPrecompileCache;
+
+    private ConcurrentDictionary<StorageCell, byte[]> _storageCache = new(LockPartitions, InitialCapacity);
+    private ConcurrentDictionary<AddressAsKey, Account> _stateCache = new(LockPartitions, InitialCapacity);
+    private ConcurrentDictionary<NodeKey, byte[]?> _rlpCache = new(LockPartitions, InitialCapacity);
+    private ConcurrentDictionary<PrecompileCacheKey, (ReadOnlyMemory<byte>, bool)> _precompileCache = new(LockPartitions, InitialCapacity);
+
+    public PreBlockCaches(ILogManager logManager)
+    {
+        _logger = logManager?.GetClassLogger<WorldState>() ?? throw new ArgumentNullException(nameof(logManager));
+        _clearStorageCache = () => _storageCache.NoResizeClear();
+        _clearStateCache = () => _stateCache.NoResizeClear();
+        _clearRlpCache = () => _rlpCache.NoResizeClear();
+        _clearPrecompileCache = () => _precompileCache.NoResizeClear();
+    }
+
+    public ConcurrentDictionary<StorageCell, byte[]> StorageCache => _storageCache;
+    public ConcurrentDictionary<AddressAsKey, Account> StateCache => _stateCache;
+    public ConcurrentDictionary<NodeKey, byte[]?> RlpCache { get; } = new(LockPartitions, InitialCapacity);
+    public ConcurrentDictionary<PrecompileCacheKey, (ReadOnlyMemory<byte>, bool)> PrecompileCache { get; } = new(LockPartitions, InitialCapacity);
 
-    public bool Clear()
+    public async Task ClearCachesInBackground()
     {
-        bool isDirty = !StorageCache.IsEmpty || !StateCache.IsEmpty || !RlpCache.IsEmpty;
+        Task t0 = Task.CompletedTask;
+        Task t1 = Task.CompletedTask;
+        Task t2 = Task.CompletedTask;
+        Task t3 = Task.CompletedTask;
+
+        bool isDirty = false;
+        if (!_storageCache.IsEmpty)
+        {
+            isDirty = true;
+            t0 = Task.Run(_clearStorageCache);
+        }
+        if (!_stateCache.IsEmpty)
+        {
+            isDirty = true;
+            t1 = Task.Run(_clearStateCache);
+        }
+        if (!_rlpCache.IsEmpty)
+        {
+            isDirty = true;
+            t2 = Task.Run(_clearRlpCache);
+        }
+        if (!_precompileCache.IsEmpty)
+        {
+            isDirty = true;
+            t3 = Task.Run(_clearPrecompileCache);
+        }
```

Comment:
put them in an array and do a loop? Will make code easier if we add more caches.

Reply:
Done